src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/DictionarySlim.cs

```diff
@@ -0,0 +1,480 @@
+
+namespace Microsoft.Collections.Extensions
+{
+    /// <summary>
+    /// DictionarySlim<TKey, TValue> is similar to Dictionary<TKey, TValue> but optimized in three ways:
+    /// 1) It allows access to the value by ref replacing the common TryGetValue and Add pattern.
+    /// 2) It does not store the hash code (assumes it is cheap to equate values).
+    /// 3) It does not accept an equality comparer (assumes Object.GetHashCode() and Object.Equals() or overridden implementation are cheap and sufficient).
+    /// </summary>
+    [DebuggerTypeProxy(typeof(Extensions.DictionarySlimDebugView<,>))]
+    [DebuggerDisplay("Count = {Count}")]
+    public class DictionarySlim<TKey, TValue> : IReadOnlyCollection<KeyValuePair<TKey, TValue>> where TKey : IEquatable<TKey>
+    {
+        // We want to initialize without allocating arrays. We also want to avoid null checks.
+        // Array.Empty would give divide by zero in modulo operation. So we use static one element arrays.
+        // The first add will cause a resize replacing these with real arrays of three elements.
+        // Arrays are wrapped in a class to avoid being duplicated for each <TKey, TValue>
+        private static readonly Entry[] InitialEntries = new Entry[1];
+        // 1-based index into _entries; 0 means empty
+        private int[] _buckets;
+        private Entry[] _entries;
+        // 0-based index into _entries of head of free chain: -1 means empty
+        private int _freeList = -1;
+
+        private struct Entry
+        {
+            public TKey key;
+            public TValue value;
+            // 0-based index of next entry in chain: -1 means end of chain
+            // also encodes whether this entry _itself_ is part of the free list by changing sign and subtracting 3,
+            // so -2 means end of free list, -3 means index 0 but on free list, -4 means index 1 but on free list, etc.
+            public int next;
+        }
+
+        public DictionarySlim()
+        {
+            _buckets = HashHelpers.DictionarySlimSizeOneIntArray;
+            _entries = InitialEntries;
+        }
+
+        public DictionarySlim(int capacity)
+        {
+            capacity = HashHelpers.GetPrime(capacity);
+            _buckets = new int[capacity];
+            _entries = new Entry[capacity];
+        }
+
+        // Drop sign bit to ensure non negative index
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private uint GetHashCode(TKey key) => (uint)key.GetHashCode();
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private int GetEntryIndex(int bucketIndex) => _buckets[bucketIndex] - 1;
+
+        public int Count { get; private set; }
+
+        public int Capacity => _entries.Length;
+
+        public bool ContainsKey(TKey key)
+        {
+            Entry[] entries = _entries;
+            int collisionCount = 0;
+            for (int i = GetEntryIndex((int)(GetHashCode(key) % (uint)entries.Length));
```

Comment:
Looks to me that it should be `% buckets.Length`, not `% entries.Length`. The result is used to index into buckets, not entries.

Unsigned `% buckets.Length` may also allow avoiding an array range check in the future.

Reply:
Right, but since they're always the same size, and _buckets is not in a local here, I guess @AnthonyLloyd was trying to avoid reading another field. I'll try with the change.

Reply:
The `_buckets` field will end up in a local variable anyway, the JIT has no other way to generate a range check safely.